File: Leetcode/5/5.cpp
class Solution {
public:
    // O(n^2) in dp
    string longestPalindrome(string s) {
        int n = (int)s.size();
        int l = -1, r = -1, maxLength = 0;
        int left, right;
        int currLength = 0;
        // odd length
        for (int i = 0; i < n; ++i) {
            left = i;
            right = i;
            do {
                currLength = right - left + 1;
                if (currLength > maxLength) {
                    maxLength = currLength;
                    l = left;
                    r = right;
                }
                --left;
                ++right;
            } while (left >= 0 && right < n && s[left] == s[right]);
        }
        // even length
        for (int i = 0; i < n-1; ++i) {
            if (s[i] == s[i+1]) {
                left = i;
                right = i+1;
                do {
                    currLength = right - left + 1;
                    if (currLength > maxLength) {
                        maxLength = currLength;
                        l = left;
                        r = right;
                    }
                    --left;
                    ++right;
                } while (left >= 0 && right < n && s[left] == s[right]);
            }
        }
        return s.substr(l, maxLength);
    }
};

File: Leetcode/5/5.go
func longestPalindrome(s string) string {
	longestLength := 0
	left, right := 0, 0
	// odd
	for i := 0; i < len(s); i++ {
		currentLength := 1
		for j, k := i-1, i+1; j >= 0 && k < len(s) && s[j] == s[k]; j, k = j-1, k+1 {
			currentLength = k - j + 1
			if currentLength > longestLength {
				longestLength = currentLength
				left = j
				right = k
			}
		}
	}
	// even
	for i := 0; i < len(s)-1; i++ {
		if s[i] == s[i+1] {
			currentLength := 2
			for j, k := i, i+1; j >= 0 && k < len(s) && s[j] == s[k]; j, k = j-1, k+1 {
				currentLength = k - j + 1
				if currentLength > longestLength {
					longestLength = currentLength
					left = j
					right = k
				}
			}
		}
	}
	return s[left:right+1]
}